tests/Progress/SnapshotTest.php

```diff
@@ -359,12 +361,223 @@ public function of_activity_runs_the_above_for_all_activity_instances()
+
+        $module1Evaluation = $this->createModuleEvaluationInstance(true, true, true, true, 100);
+
+        $moduleInstanceEvaluator = $this->prophesize(ModuleInstanceEvaluator::class);
+        $this->useEvaluationForModuleInstance($moduleInstanceEvaluator, $activityInstance, $moduleInstance, $module1Evaluation);
+        \BristolSU\Support\ModuleInstance\Facade\ModuleInstanceEvaluator::swap($moduleInstanceEvaluator->reveal());
+
+        $progressUpdateRepository = $this->prophesize(ProgressUpdateRepository::class);
+
+        $progressUpdateRepository->hasChanged(
+            $activityInstance->id,
+            'my-caller',
+            Argument::that(function($progress) use ($module1Progress) {
+                return $progress->getModules()[0] != $module1Progress;
+            }))->willReturn(true);
+
+        $progressUpdateRepository->saveChanges(
+            $activityInstance->id,
+            'my-caller',
+            Argument::that(function($progress) use ($module1Progress) {
+                return $progress->getModules()[0] != $module1Progress;
+            })
+        )->shouldBeCalled();
+
+        $snapshot = new Snapshot($progressUpdateRepository->reveal());
+
+        $this->assertNotNull(
+            $snapshot->ofUpdateToActivityInstance($activityInstance, 'my-caller')
+        );
+    }
+
+    /** @test */
+    public function ofUpdatesToActivity_returns_an_empty_array_if_no_progress_has_changed(){
+        Carbon::setTestNow(Carbon::now());
+        $activity = factory(Activity::class)->create();
+        $moduleInstance = factory(ModuleInstance::class)->create(['activity_id' => $activity->id]);
+        $activityInstance = factory(ActivityInstance::class)->create(['activity_id' => $activity->id]);
+
+        $module1Progress = $this->createModuleInstanceProgress($moduleInstance->id, true, true, true, true, 100);
+
+        $module1Evaluation = $this->createModuleEvaluationInstance(true, true, true, true, 100);
+
+        $moduleInstanceEvaluator = $this->prophesize(ModuleInstanceEvaluator::class);
+        $this->useEvaluationForModuleInstance($moduleInstanceEvaluator, $activityInstance, $moduleInstance, $module1Evaluation);
+        \BristolSU\Support\ModuleInstance\Facade\ModuleInstanceEvaluator::swap($moduleInstanceEvaluator->reveal());
+
+        $progressUpdateRepository = $this->prophesize(ProgressUpdateRepository::class);
+
+        $progressUpdateRepository->hasChanged(
+            $activityInstance->id,
+            'my-caller',
+            Argument::that(function($progress) use ($module1Progress) {
+                return $progress->getModules()[0] == $module1Progress;
+            }))->willReturn(false);
+
+        $snapshot = new Snapshot($progressUpdateRepository->reveal());
+
+        $this->assertEmpty(
+            $snapshot->ofUpdatesToActivity($activity, 'my-caller')
+        );
+    }
+
+    /** @test */
+    public function ofUpdatesToActivity_returns_only_changed_progresses(){
```

Comment:
For this test, you should set up 2 or 3 module progress/evaluations. Then check that, if only two of them change (e.g. hasChanged is true for tw of the progresses, false for one), that only those two progresses are changed! 

If you take away the `array_filter` function in Snapshot, you expect this test to fail because it should only return changed progresses. Instead, however, it'll just return the two changed progresses and a null value, which currently lets the test pass! Therefore, it isn't quite testing enough of what it needs to